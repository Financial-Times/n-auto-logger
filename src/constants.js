export const CATEGORIES = {
	FETCH_RESPONSE_OK: 'FETCH_RESPONSE_OK',
	FETCH_RESPONSE_ERROR: 'FETCH_RESPONSE_ERROR',
	FETCH_NETWORK_ERROR: 'FETCH_NETWORK_ERROR',
};

export default {
	CATEGORIES,
};
