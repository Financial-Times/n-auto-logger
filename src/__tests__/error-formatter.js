import 'isomorphic-fetch';
import nodeFetch from 'node-fetch';
import {
	formatFetchResponseError,
	formatFetchNetworkError,
	formatFetchError,
} from '../error-formatter';
import { CATEGORIES } from '../constants';

describe('formatFetchResponseError', () => {
	it('reports wrong response being threw to catch if response is ok', async () => {
		const headers = new Headers();
		headers.append('content-type', 'text/plain; charset=utf-8');
		const response = new Response('', { status: 200, headers });
		const formatted = await formatFetchResponseError(response);
		expect(formatted.category).toBe(CATEGORIES.FETCH_RESPONSE_OK);
		expect(formatted).toMatchSnapshot();
	});

	it('format error in text/html contentType correctly', async () => {
		const headers = new Headers();
		headers.append('content-type', 'text/html; charset=utf-8');
		const message = '<html></html>';
		const e = new Response(message, { status: 404, headers });
		const formatted = await formatFetchResponseError(e);
		expect(formatted.message).toBe(message);
		expect(formatted).toMatchSnapshot();
	});

	it('format error in text/plain contentType correctly', async () => {
		const headers = new Headers();
		headers.append('content-type', 'text/plain; charset=utf-8');
		const message = '403 Forbidden';
		const e = new Response(message, { status: 403, headers });
		const formatted = await formatFetchResponseError(e);
		expect(formatted.message).toBe(message);
		expect(formatted).toMatchSnapshot();
	});

	it('format error in application/json contentType correctly', async () => {
		const headers = new Headers();
		headers.append('content-type', 'application/json; charset=utf-8');
		const message = { message: 'some message', document: 'some url' };
		const e = new Response(JSON.stringify(message), { status: 404, headers });
		const formatted = await formatFetchResponseError(e);
		expect(formatted.message).toEqual(message);
		expect(formatted).toMatchSnapshot();
	});
});

describe('formatFetchNetworkError', () => {
	it('format network error correctly', async () => {
		const e = new nodeFetch.FetchError(
			'request to https://mock.com/ failed, reason: unable to verify the first certificate',
		);
		e.code = 'UNABLE_TO_VERIFY_LEAF_SIGNATURE';
		const formatted = await formatFetchNetworkError(e);
		expect(formatted.category).toBe(CATEGORIES.FETCH_NETWORK_ERROR);
		expect(formatted).toMatchSnapshot();
	});
});

describe('formatFetchError', () => {
	it('format node-fetch response error correctly', async () => {
		const headers = new nodeFetch.Headers();
		headers.append('content-type', 'text/plain; charset=utf-8');
		const e = new nodeFetch.Response('403 Forbidden', { status: 403, headers });
		const formatted = await formatFetchError(e);
		expect(formatted.category).toBe(CATEGORIES.FETCH_RESPONSE_ERROR);
		expect(formatted).toMatchSnapshot();
	});

	it('format isomorphic-fetch response error correctly', async () => {
		const headers = new Headers();
		headers.append('content-type', 'text/plain; charset=utf-8');
		const e = new Response('403 Forbidden', { status: 403, headers });
		const formatted = await formatFetchError(e);
		expect(formatted.category).toBe(CATEGORIES.FETCH_RESPONSE_ERROR);
		expect(formatted).toMatchSnapshot();
	});

	it('format network error correctly', async () => {
		const e = new nodeFetch.FetchError(
			'request to https://mock.com/ failed, reason: unable to verify the first certificate',
		);
		e.code = 'UNABLE_TO_VERIFY_LEAF_SIGNATURE';
		const formatted = await formatFetchError(e);
		expect(formatted.category).toBe(CATEGORIES.FETCH_NETWORK_ERROR);
		expect(formatted).toMatchSnapshot();
	});

	it('does nothing on other error types e.g. System Error ', async () => {
		const e = new Error('some error message');
		const formatted = await formatFetchError(e);
		expect(formatted).toBe(e);
	});
});
