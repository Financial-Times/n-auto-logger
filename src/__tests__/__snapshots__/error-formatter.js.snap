// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatFetchError format isomorphic-fetch response error correctly 1`] = `
Object {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "text/plain; charset=utf-8",
  "message": "403 Forbidden",
  "status": 403,
}
`;

exports[`formatFetchError format network error correctly 1`] = `
Object {
  "category": "FETCH_NETWORK_ERROR",
  "code": "UNABLE_TO_VERIFY_LEAF_SIGNATURE",
  "message": "request to https://mock.com/ failed, reason: unable to verify the first certificate",
}
`;

exports[`formatFetchError format node-fetch response error correctly 1`] = `
Object {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "text/plain; charset=utf-8",
  "message": "403 Forbidden",
  "status": 403,
}
`;

exports[`formatFetchNetworkError format network error correctly 1`] = `
Object {
  "category": "FETCH_NETWORK_ERROR",
  "code": "UNABLE_TO_VERIFY_LEAF_SIGNATURE",
  "message": "request to https://mock.com/ failed, reason: unable to verify the first certificate",
}
`;

exports[`formatFetchResponseError format error in application/json contentType correctly 1`] = `
Object {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "application/json; charset=utf-8",
  "message": Object {
    "document": "some url",
    "message": "some message",
  },
  "status": 404,
}
`;

exports[`formatFetchResponseError format error in text/html contentType correctly 1`] = `
Object {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "text/html; charset=utf-8",
  "message": "<html></html>",
  "status": 404,
}
`;

exports[`formatFetchResponseError format error in text/plain contentType correctly 1`] = `
Object {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "text/plain; charset=utf-8",
  "message": "403 Forbidden",
  "status": 403,
}
`;

exports[`formatFetchResponseError reports wrong response being threw to catch if response is ok 1`] = `
Object {
  "category": "FETCH_RESPONSE_OK",
  "message": "it shouldn't be caught as exception, please check the code",
  "status": 200,
}
`;
